rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -0,0 +1,613 @@
+  std::string service_type = request_type_support->get_name();
+  size_t suffix_substring_position = service_type.find("Request_");
+  if (std::string::npos != suffix_substring_position) {
+    service_type = service_type.substr(0, suffix_substring_position);
+  } else {
+    RMW_ZENOH_LOG_ERROR_NAMED(
+      "rmw_zenoh_cpp",
+      "Unexpected type %s for client %s. Report this bug",
+      service_type.c_str(), service_name.c_str());
+    return nullptr;
+  }
+
+  // Convert the type hash to a string so that it can be included in the keyexpr.
+  char * type_hash_c_str = nullptr;
+  rcutils_ret_t stringify_ret = rosidl_stringify_type_hash(
+    type_hash,
+    *allocator,
+    &type_hash_c_str);
+  if (RCUTILS_RET_BAD_ALLOC == stringify_ret) {
+    RMW_SET_ERROR_MSG("Failed to allocate type_hash_c_str.");
+    return nullptr;
+  }
+  auto free_type_hash_c_str = rcpputils::make_scope_exit(
+    [&allocator, &type_hash_c_str]() {
+      allocator->deallocate(type_hash_c_str, allocator->state);
+    });
+
+  std::size_t domain_id = node_info.domain_id_;
+  auto entity = liveliness::Entity::make(
+    z_info_zid(session),
+    std::to_string(node_id),
+    std::to_string(service_id),
+    liveliness::EntityType::Client,
+    std::move(node_info),
+    liveliness::TopicInfo{
+      std::move(domain_id),
+      service_name,
+      std::move(service_type),
+      type_hash_c_str,
+      std::move(adapted_qos_profile)}
+  );
+  if (entity == nullptr) {
+    RMW_ZENOH_LOG_ERROR_NAMED(
+      "rmw_zenoh_cpp",
+      "Unable to generate keyexpr for liveliness token for the client %s.",
+      service_name);
+    return nullptr;
+  }
+
+  auto client_data = std::shared_ptr<ClientData>(
+    new ClientData{
+      node,
+      std::move(entity),
+      request_members,
+      response_members,
+      std::move(request_type_support),
+      std::move(response_type_support)
+    });
+
+  client_data->keyexpr_ =
+    z_keyexpr_new(client_data->entity_->topic_info().value().topic_keyexpr_.c_str());
+  auto free_ros_keyexpr = rcpputils::make_scope_exit(
+    [client_data]() {
+      z_drop(z_move(client_data->keyexpr_));
+    });
+  if (!z_check(z_loan(client_data->keyexpr_))) {
+    RMW_SET_ERROR_MSG("unable to create zenoh keyexpr.");
+    return nullptr;
+  }
+
+  client_data->token_ = zc_liveliness_declare_token(
+    session,
+    z_keyexpr(client_data->entity_->liveliness_keyexpr().c_str()),
+    NULL
+  );
+  auto free_token = rcpputils::make_scope_exit(
+    [client_data]() {
+      z_drop(z_move(client_data->token_));
+    });
+  if (!z_check(client_data->token_)) {
+    RMW_ZENOH_LOG_ERROR_NAMED(
+      "rmw_zenoh_cpp",
+      "Unable to create liveliness token for the client.");
+    return nullptr;
+  }
+
+  free_ros_keyexpr.cancel();
+  free_token.cancel();
+
+  // Erase from deleted_clients set if the memory address for the client data is reused.
+  num_in_flight_map.erase(client_data.get());
+  deleted_clients.erase(client_data.get());
```

Comment:
There are a few different problems with this that I see.

First, this needs to take the `num_in_flight_mutex` for mutual exclusion.

Second, I don't think that we ever need to erase from the `num_in_flight_map`.  That only exists for the lifetime of a `ClientData` object.  In point of fact, we may actually want to add this object to the map right here (with a `num_in_flight` of 0).

Finally, I don't think that erasing this from the `deleted_clients` is the right thing to do.  If we take my other suggestion on only adding to the `deleted_clients` map if queries are in-flight, then removing this here means that those outstanding queries may do the wrong thing during `client_data_handler`.  That is, they will see that they are *not* in the `deleted_clients` map, and go on to operate on a different object.  This is actually tricky to fix, though I have a somewhat hacky idea.  That idea is to continually allocate `ClientData` structures (as in line 236) until we find an address that is *not* in `deleted_clients`.  Once we find one, we deallocate all of the other ones.  That way we'll be guaranteed to get an address that does not have queries currently in flight.

Reply:
> That idea is to continually allocate ClientData structures (as in line 236) until we find an address that is not in deleted_clients. Once we find one, we deallocate all of the other ones. That way we'll be guaranteed to get an address that does not have queries currently in flight.

That should resolve it but wondering we need to go through this hassle now especially since we will undo all these changes once switch to zenoh-cpp and will rely on lambda captures?

Reply:
Personally, I think it is worth it.  It's not a whole lot of code, and we never know what is going to happen in the future.  The good news is that it is not too hard; see #304, which targets this PR and fixes all of these issues (I think).